### PR TITLE
Remove all #[inline] attributes

### PR DESCRIPTION
**Pull request by k-sareen**

### CODE REVIEW

**Comment by qinsoon** [comment]
According to the results we have (https://mmtk.zulipchat.com/#narrow/stream/262673-mmtk-core/topic/inline.20directives/near/321556606), simply removing all the inline directives will slow down our GC by a few percentages. And its performance impact on the mutator time is unknown (we measured with a JIT-generated allocation fastpath). 

Our goal on https://github.com/mmtk/mmtk-core/issues/733 is to improve performance (or at least not degrade performance), and remove as many inline directives as possible (so that we can have a cleaner code base). Simply removing all the inline directives won't achieve the goal.

**Comment by k-sareen** [comment]
Technically if the PGO build works, then we never have to touch any `inline` directives at all. So I guess we should merge this PR after we know PGO works.

**Comment by k-sareen** [comment]
This commit is stale so don't merge it yet. I'll add a script that directly builds the PGO build as well. Though maybe that should sit in the binding repo?

**Comment by k-sareen** [comment]
@qinsoon How should I lay out the two PRs? Should I make a PGO section in the READMEs of both mmtk-core and mmtk-openjdk?

**Comment by qinsoon** [comment]
> @qinsoon How should I lay out the two PRs? Should I make a PGO section in the READMEs of both mmtk-core and mmtk-openjdk?

This is just what I am thinking about.

For mmtk-core, you can 
1. add a subsection in the Build section in mmtk-core, mentioning that we recommend using PGO when you try build and measure performance with MMTk. You can just include a link to the Rust PGO guide. 
2. add some comments in the `lib.rs` to mention that we deliberately remove all the inline directives, and rely on the compiler and PGO for inlining.

For the OpenJDK binding, you can include a more detailed guide about how to build the OpenJDK binding with PGO.

**Comment by k-sareen** [comment]
Alright makes sense to me. I'll do that. Note that I've also bumped the Rust version to 1.66.1 (due to LLVM tool compatibility) in this PR. That's why we're seeing some new style checks I presume.

**Comment by qinsoon** [comment]
> Alright makes sense to me. I'll do that. Note that I've also bumped the Rust version to 1.66.1 (due to LLVM tool compatibility) in this PR. That's why we're seeing some new style checks I presume.

I would suggest not upgrading the Rust toolchain in this PR. I assume you do not have to mention the exact version in the build description.

**Comment by k-sareen** [comment]
Right -- the results above were based on Rust v1.66.1. So we should just keep that in mind. I think we have `llvm-tools` installed for the older Rust version as well.

**Comment by qinsoon** [comment]
> Right -- the results above were based on Rust v1.66.1. So we should just keep that in mind. I think we have `llvm-tools` installed for the older Rust version as well.
> 
> EDIT: @qinsoon Actually no. We don't have the `llvm-tools` installed for the older Rust version. Only for v1.66.1.

OK. Then we probably have to upgrade the Rust version.

**Comment by k-sareen** [comment]
So the previous result mistakenly compiled `master` with Rust v1.59.0 as opposed to Rust v1.66.1. This result compiles `master` with Rust v1.66.1: http://squirrel.anu.edu.au/plotty/kunals/bench-test/p/QTMP3M. GenImmix and Immix see performance benefits (or at least no performance regressions), but there is a 10% slowdown for SemiSpace.

As an aside, changing Rust version to v1.66.1 has a big benefit as we can see in this result: http://squirrel.anu.edu.au/plotty/kunals/bench-test/p/WDptqC

**Comment by k-sareen** [comment]
@qinsoon I don't know if I should fix all the clippy errors in this PR. It's unrelated to remove `inline` attributes. I think the best way to approach this is to have a PR that changes Rust version to v1.66.1 that will fix all the clippy errors as well and then we merge this PR after that (same for the mmtk-openjdk PR). Does that make sense to you?

**Comment by qinsoon** [comment]
> @qinsoon I don't know if I should fix all the clippy errors in this PR. It's unrelated to remove `inline` attributes. I think the best way to approach this is to have another PR that changes Rust version to v1.66.1 that will fix all the clippy errors as well and then we merge this PR after that (same for the mmtk-openjdk PR). Does that make sense to you?

Yes. I agree.

File: README.md
Comment (qinsoon) [inline README.md:52]: This is not true according to your results. You can update this section once we have a conclusion.

**Review by qinsoon — APPROVED** [review]
LGTM

File: README.md
Comment (k-sareen) [reply README.md:52]: Yeah -- for the time-being it's fine, I think. I can edit it to say something like "We use our best-performing GC for the profiling step in order to get the best performance results", if you want.

**Pull request merged**
